cmd/rpcdaemon/commands/erigon_block.go

```diff
@@ -58,3 +58,72 @@ func (api *ErigonImpl) GetHeaderByHash(ctx context.Context, hash common.Hash) (*
 
 	return header, nil
 }
+
+func (api *ErigonImpl) GetBlockByTimeStamp(ctx context.Context, timeStamp uint64) (types.Block, error) {
+	tx, err := api.db.BeginRo(ctx)
+	if err != nil {
+		return types.Block{}, err
+	}
+	defer tx.Rollback()
+
+	currentHeader := rawdb.ReadCurrentHeader(tx)
+	currenttHeaderTime := currentHeader.Time
+
+	var lowestNumber uint64 = 1
+	highestNumber := currentHeader.Number.Uint64()
+	middleNumber := (highestNumber + lowestNumber) / 2
+
+	firstHeader := rawdb.ReadHeaderByNumber(tx, lowestNumber)
+	firstHeaderTime := firstHeader.Time
+
+	middleHeader := rawdb.ReadHeaderByNumber(tx, middleNumber)
+
+	if currenttHeaderTime == timeStamp {
+		block := rawdb.ReadCurrentBlock(tx)
+		return *block, nil
+	}
+
+	if firstHeaderTime == timeStamp {
+		block, err := rawdb.ReadBlockByNumber(tx, lowestNumber)
+		if err != nil {
+			return types.Block{}, err
+		}
+		return *block, nil
+	}
+
+	if middleHeader.Time == timeStamp {
+		block, err := rawdb.ReadBlockByNumber(tx, middleNumber)
+		if err != nil {
+			return types.Block{}, err
+		}
+
+		return *block, nil
+	}
+
+	for lowestNumber < highestNumber {
+
+		if middleHeader.Time < timeStamp {
+			lowestNumber = middleNumber
+		}
+
+		if middleHeader.Time > timeStamp {
+			highestNumber = middleNumber
+		}
+
+		if middleHeader.Time == timeStamp {
+			block, err := rawdb.ReadBlockByNumber(tx, middleNumber)
+			if err != nil {
+				return types.Block{}, err
+			}
+
+			return *block, nil
+		}
+
+		middleNumber = (highestNumber + lowestNumber) / 2
+		middleHeader = rawdb.ReadHeaderByNumber(tx, middleNumber)
+
+	}
+
+	return types.Block{}, nil
```

Comment:
You should probably return the block with `highestNumber`. so in the case where precise timestamp is not found, you return the block just after that timestamp